### PR TITLE
Split `IGitRepo` and `ILocalGitRepo`, add submodule listing

### PR DESCRIPTION
**Pull request by premun**
This is another attempt at refactoring out the `ILocalGitRepo` interface after the previous change was reverted (https://github.com/dotnet/arcade-services/pull/2030).

This change is much less invasive and tackles the problem in a better way than before. I have already ran the scenario tests locally and they pass.

#### This change:
- is a first step in untangling the `Remote` class and removes following inheritance `IGitRepo : ILocalGitRepo`,
- adds the ability to use and extend `ILocalGitRepo` with functionality tied to local `.git` folders which will be needed in the VMR later,
- adds the capability to list git submodules to `ILocalGitRepo`,
- removes some dead code.

#### TLDR
This change removes following inheritance `IGitRepo : ILocalGitRepo`. It then splits `IGitRepo` into `IGitRemoteRepo` and `IGitRepo`. `IGitRepo` is then re-used in `ILocalGitRepo`.
The outcome is that `GitHubClient` and `AzDOGitClient` implement `IGitRemoteRepo` which also makes more sense from the naming perspective.

#### Problems this solves:
- Many implementations of `IGitRepo` that had to implement `ILocalGitRepo` too had `throw new NotImplemented..`.
- The only method used from `ILocalGitRepo` via `IGitRepo` was `GetFilesAtCommitAsync` whose implementation in `ILocalGitRepo` didn't make sense - it accepted URI and SHA but didn't use either, just listed files in a folder. It only was used in one place though `Local.cs` so I moved it there
- I removed a lot of dead code (`ILocalGitRepo` methods on `IGitRepo` classes that are not used).
- I removed some dead code in unit tests which was setting up things that were not used in the tests.

The largest weirdness that I haven't dealt with so far comes from `RemoteFactory` implementations where we supply `null` instead of `IGitRepo` often. In those cases, we supply the `ILocalGitRepo` as `null` too, not breaking anything. For instance, we have things like "BAR only client" that only has the `BarClient` set inside `Remote` and this will obviously break if we try to invoke any of the non-bar methods but this is not changed in this PR.

### CODE REVIEW

File: src/Microsoft.DotNet.Darc/src/DarcLib/Actions/Local.cs [hunk 1]
```diff
@@ -165,5 +166,17 @@ public void AddRemoteIfMissing(string repoDir, string repoUrl)
         {
             _gitClient.AddRemoteIfMissing(repoDir, repoUrl);
         }
+
+        private List<GitFile> GetFilesAtRelativeRepoPathAsync(string path)
```
Comment (tkapin) [inline src/Microsoft.DotNet.Darc/src/DarcLib/Actions/Local.cs:170]: It'd be good to have this method covered by unit tests

**Review by tkapin — APPROVED** [review]

File: src/Microsoft.DotNet.Darc/src/DarcLib/Actions/Local.cs (hunk 1 above)
Comment (premun) [reply src/Microsoft.DotNet.Darc/src/DarcLib/Actions/Local.cs:170]: It was just moved in this PR, I'd leave adding tests for the debt epic honestly
